espnet2/enh/espnet_model.py

```diff
@@ -215,488 +132,52 @@ def forward(
-                        others["mask_dereverb{}".format(spk + 1)]
-                        for spk in range(self.num_spk)
-                        if "mask_dereverb{}".format(spk + 1) in others
-                    ]
-                    assert len(mask_wpe_pre) == dereverb_speech_ref.size(1), (
-                        len(mask_wpe_pre),
-                        dereverb_speech_ref.size(1),
-                    )
-                    dereverb_speech_ref = torch.unbind(dereverb_speech_ref, dim=1)
-                    dereverb_spectrum_ref = [
-                        self.encoder(dr, speech_lengths)[0]
-                        for dr in dereverb_speech_ref
-                    ]
-                    dereverb_mask_ref = self._create_mask_label(
-                        spectrum_mix, dereverb_spectrum_ref, mask_type=self.mask_type
-                    )
-
-                    tf_dereverb_loss, perm_d = self._permutation_loss(
-                        dereverb_mask_ref, mask_wpe_pre, loss_func
-                    )
-                    tf_loss = tf_loss + tf_dereverb_loss
-
-                if "mask_noise1" in others:
-                    if noise_ref is None:
-                        raise ValueError(
-                            "No noise reference for training!\n"
-                            'Please specify "--use_noise_ref true" in run.sh'
-                        )
-
-                    noise_ref = torch.unbind(noise_ref, dim=1)
-                    noise_spectrum_ref = [
-                        self.encoder(nr, speech_lengths)[0] for nr in noise_ref
-                    ]
-                    noise_mask_ref = self._create_mask_label(
-                        spectrum_mix, noise_spectrum_ref, mask_type=self.mask_type
-                    )
-
-                    mask_noise_pre = [
-                        others["mask_noise{}".format(n + 1)]
-                        for n in range(self.num_noise_type)
-                    ]
-                    tf_noise_loss, perm_n = self._permutation_loss(
-                        noise_mask_ref, mask_noise_pre, loss_func
-                    )
-                    tf_loss = tf_loss + tf_noise_loss
-            else:
-                raise ValueError("Unsupported loss type: %s" % self.loss_type)
-
-            loss = tf_loss
-            return loss, spectrum_pre, others, flens, perm
-
-        else:
+        if feature_pre is not None:
             speech_pre = [self.decoder(ps, speech_lengths)[0] for ps in feature_pre]
-            if not cal_loss:
-                loss, perm = None, None
-                return loss, speech_pre, None, speech_lengths, perm
-
-            # speech_pre: list[(batch, sample)]
-            assert speech_pre[0].dim() == 2, speech_pre[0].dim()
-
-            if speech_ref.dim() == 4:
-                # For si_snr loss of multi-channel input,
-                # only select one channel as the reference
-                speech_ref = speech_ref[..., self.ref_channel]
```

Comment:
We should reuse this part because sometimes the reference can be multi-channel while the network output is single-channel.

Reply:
Done.